### PR TITLE
Adds redux code for displaying shared hparam columns in the Time Series page

### PR DESCRIPTION
**Pull request by hoonji**
## Motivation for features / changes
Required redux edits for displaying hparams in time series.

## Technical description of changes
Adds a new property to HparamsState, `dashboardDisplayedHparamColumns` to represent the columns shown across the time series page (similar in concept to the existing `dashboardFilters`)

Adds four actions for hparam column manipulation:
- dashboardHparamColumnAdded
- dashboardHparamColumnRemoved
- dashboardHparamColumnToggled
- dashboardHparamColumnOrderChanged

and reducer logic to handle these events.

Note that the add actions cannot simply pass the header index in this case as the table components will not have knowledge of the relative index of hparam columns. The actions instead specify the column to insert next to and the side, and the indices are computed from within the reducer (this can be seen as simply moving the index computation from the component to the reducer, which better conforms to https://redux.js.org/style-guide/#put-as-much-logic-as-possible-in-reducers )

Also exports required types from widgets/data_table - these will be also used by the runs and metrics tables the following PRs

## Detailed steps to verify changes work correctly (as executed by you)
- Unit tests pass

## Alternate designs / implementations considered (or N/A)
An alternative is to duplicate the hparams state across the runs and scalar card tables as RunsDataState.runsTableHeaders and MetricsState.singleSelectionHeaders/MetricsState.rangeSelectionHeaders, but this would either incur massive complications in syncing the various column operations (add/remove/hide/sort) or alternatively, would force the user to duplicate operations multiple times to get the hparams columns in sync. The chosen design also aligns with [keeping state minimal](https://redux.js.org/style-guide/#keep-state-minimal-and-derive-additional-values).

### CODE REVIEW

File: tensorboard/webapp/hparams/_redux/hparams_reducers_test.ts [hunk 1]
```diff
@@ -309,4 +310,424 @@ describe('hparams/_redux/hparams_reducers_test', () => {
...
+      },
+      {
+        type: ColumnHeaderType.HPARAM,
+        name: 'conv_kernel_size',
+        displayName: 'Conv Kernel Size',
+        enabled: true,
+      },
+    ];
+
+    it('appends an hparam column to the end', () => {
```
Comment (hoonji) [inline tensorboard/webapp/hparams/_redux/hparams_reducers_test.ts:330]: QQ: is there a recommended way to do parameterized testing in 3p code? Would something like the example in https://stackoverflow.com/questions/38659880/jasmine-parameterized-unit-test be acceptable? I haven't found any such examples in the codebase so far

Namely:
```
it('action(value) should reset the forms pool only if value is true', () => {
  [
    { value: true, calledTimes: 1 },
    { value: false, calledTimes: 0 },
  ].forEach(({ value, calledTimes }) => {
    spyResetFormsPool.calls.reset();

    component.action(value);

    expect(spyResetFormsPool).toHaveBeenCalledTimes(calledTimes);
  });
});
```

File: tensorboard/webapp/hparams/_redux/hparams_reducers.ts [hunk 2]
```diff
@@ -87,7 +89,82 @@ const reducer: ActionReducer<HparamsState, Action> = createReducer(
...
+    if (nextTo !== undefined && side !== undefined) {
+      const nextToIndex = oldColumns.findIndex(
+        (col) => col.name === nextTo.name
+      );
+      if (nextToIndex !== -1) {
+        destinationIndex = side === Side.RIGHT ? nextToIndex + 1 : nextToIndex;
+      }
+    }
+    let newColumn = {...column, enabled: true};
+    let newColumns = [...oldColumns];
```
Comment (rileyajones) [inline tensorboard/webapp/hparams/_redux/hparams_reducers.ts:106]: This is a little nitpicky but I definitely prefer to keep as variables `const` when possible.
```suggestion
    const newColumn = {...column, enabled: true};
    const newColumns = [...oldColumns];
```

File: tensorboard/webapp/hparams/_redux/hparams_reducers.ts [hunk 3]
```diff
@@ -87,7 +89,82 @@ const reducer: ActionReducer<HparamsState, Action> = createReducer(
...
+  on(actions.dashboardHparamColumnToggled, (state, {column: toggledColumn}) => {
+    const newColumns = state.dashboardDisplayedHparamColumns.map((column) => {
+      if (column.name === toggledColumn.name) {
+        return {
+          ...column,
+          enabled: !toggledColumn.enabled,
+        };
+      } else {
+        return column;
+      }
```
Comment (rileyajones) [inline tensorboard/webapp/hparams/_redux/hparams_reducers.ts:133]: nit: the else clause here is redundant
```suggestion
      }
      return column;
```

File: tensorboard/webapp/hparams/_redux/hparams_reducers_test.ts (hunk 1 above)
Comment (rileyajones) [reply tensorboard/webapp/hparams/_redux/hparams_reducers_test.ts:330]: You should be able to call `it` in a loop.
```
[ { value: true, calledTimes: 1 }, { value: false, calledTimes: 0 }, ].forEach(({ value, calledTimes }) => {
  it(`action(${value}) should reset the forms pool only if value is true`, () => {
   spyResetFormsPool.calls.reset(); component.action(value); expect(spyResetFormsPool).toHaveBeenCalledTimes(calledTimes);
  });
```

File: tensorboard/webapp/hparams/_redux/hparams_reducers.ts [hunk 4]
```diff
@@ -87,7 +89,82 @@ const reducer: ActionReducer<HparamsState, Action> = createReducer(
...
+    return {
+      ...state,
+      dashboardDisplayedHparamColumns: newColumns,
+    };
+  }),
+  on(
+    actions.dashboardHparamColumnOrderChanged,
+    (state, {source, destination, side}) => {
+      const {dashboardDisplayedHparamColumns: columns} = state;
+      let sourceIndex = columns.findIndex(
```
Comment (rileyajones) [inline tensorboard/webapp/hparams/_redux/hparams_reducers.ts:145]: nit 
```suggestion
      const sourceIndex = columns.findIndex(
```

File: tensorboard/webapp/hparams/_redux/hparams_reducers_test.ts [hunk 5]
```diff
@@ -309,4 +310,424 @@ describe('hparams/_redux/hparams_reducers_test', () => {
       });
     });
   });
+
+  describe('dashboardHparamColumnAdded', () => {
+    const fakeColumns = [
```
Comment (rileyajones) [inline tensorboard/webapp/hparams/_redux/hparams_reducers_test.ts:315]: Could you please add a couple more columns here?

File: tensorboard/webapp/hparams/_redux/hparams_reducers.ts [hunk 6]
```diff
@@ -87,7 +89,82 @@ const reducer: ActionReducer<HparamsState, Action> = createReducer(
...
+      if (sourceIndex === -1 || sourceIndex === destinationIndex) {
+        return state;
+      }
+      if (destinationIndex === -1) {
+        destinationIndex = side === Side.LEFT ? 0 : columns.length - 1;
+      }
+
+      const newColumns = [...columns];
+      newColumns.splice(sourceIndex, 1);
+      newColumns.splice(destinationIndex, 0, source);
```
Comment (rileyajones) [inline tensorboard/webapp/hparams/_redux/hparams_reducers.ts:160]: If the destination IS found, the `side` is not taken into account.

Heads up, if `side` IS accounted for you will likely run into an issue with way you derive the indexes, then mutate the array.

Comment (hoonji) [reply tensorboard/webapp/hparams/_redux/hparams_reducers.ts:160]: > If the destination IS found, the side is not taken into account.

Yeah this is WAI: in typical cases, we just swap source with dest - no Side required. Side is essentially used as a backup for when dest is not found. This allows us to do things like: reorder

I'll add additional conditional logic to make sure this only triggers when dest is not found AND side exists, and will add a comment to clarify this behavior.

PLMK if this seems unnecessarily convoluted - I'm also open to simply removing the side param and simply doing nothing if dest isn't found. (IMO reordering elements to the end of the "column group" should be rather intuitive though)

File: tensorboard/webapp/hparams/_redux/hparams_reducers_test.ts (hunk 5 above)
Comment (hoonji) [reply tensorboard/webapp/hparams/_redux/hparams_reducers_test.ts:315]: Sure thing, there are now 4.

File: tensorboard/webapp/hparams/_redux/hparams_reducers_test.ts (hunk 1 above)
Comment (hoonji) [reply tensorboard/webapp/hparams/_redux/hparams_reducers_test.ts:342]: Thanks! I applied this pattern where tests were equivalent except for the side parameter

File: tensorboard/webapp/hparams/_redux/hparams_reducers.ts (hunk 2 above)
Comment (hoonji) [reply tensorboard/webapp/hparams/_redux/hparams_reducers.ts]: Oops forgot to convert these. Applied!

File: tensorboard/webapp/hparams/_redux/hparams_reducers.ts (hunk 3 above)
Comment (hoonji) [reply tensorboard/webapp/hparams/_redux/hparams_reducers.ts]: Done

File: tensorboard/webapp/hparams/_redux/hparams_reducers.ts (hunk 4 above)
Comment (hoonji) [reply tensorboard/webapp/hparams/_redux/hparams_reducers.ts]: Done

**Review by rileyajones — APPROVED** [review]

**Pull request merged**